### PR TITLE
Support STARTS_WITH in visibility queries

### PR DESCRIPTION
**Pull request by rodrigozhou**
**What changed?**
Parse `STARTS_WITH` queries to visibility, and build queries to Elasticsearch and SQL.

**Why?**
Support prefix search in visibility queries.

**How did you test it?**
Added unit tests, and integration tests.

**Potential risks**
No.

**Is hotfix candidate?**
No.

### CODE REVIEW

File: common/persistence/visibility/store/query/converter.go [hunk 1]
```diff
@@ -479,24 +479,36 @@ func (c *comparisonExprConverter) Convert(expr sqlparser.Expr) (elastic.Query, e
...
+		if !ok {
+			return nil, NewConverterError("right-hand side of '%v' must be a string", comparisonExpr.Operator)
+		}
+		query = elastic.NewPrefixQuery(colName, v)
+	case sqlparser.NotStartsWithStr:
+		v, ok := colValues[0].(string)
+		if !ok {
+			return nil, NewConverterError("right-hand side of '%v' must be a string", comparisonExpr.Operator)
+		}
+		query = elastic.NewBoolQuery().MustNot(elastic.NewPrefixQuery(colName, v))
```
Comment (alexshtin) [inline common/persistence/visibility/store/query/converter.go:511]: Don't you need to enable [index_prefixes](https://www.elastic.co/guide/en/elasticsearch/reference/current/index-prefixes.html) for this to work?

File: common/persistence/visibility/store/query/converter.go [hunk 2]
```diff
@@ -479,24 +479,36 @@ func (c *comparisonExprConverter) Convert(expr sqlparser.Expr) (elastic.Query, e
...
+	case sqlparser.LessEqualStr:
 		query = elastic.NewRangeQuery(colName).Lte(colValues[0])
-	case ">":
+	case sqlparser.GreaterThanStr:
 		query = elastic.NewRangeQuery(colName).Gt(colValues[0])
-	case "<":
+	case sqlparser.LessThanStr:
 		query = elastic.NewRangeQuery(colName).Lt(colValues[0])
-	case "=", "like": // The only difference is that "%" is removed for LIKE queries.
+	case sqlparser.EqualStr, sqlparser.LikeStr: // The only difference is that "%" is removed for LIKE queries.
```
Comment (alexshtin) [inline common/persistence/visibility/store/query/converter.go:490]: If `START_WITH` is just an alias to `LIKE` you should allow it too and make them work same way.
Let's not just remove `%` from the operand but fail query if `%` is anywhere but lust symbol.

Comment (rodrigozhou) [reply common/persistence/visibility/store/query/converter.go:490]: I will remove the LIKE support completely soon.

File: common/persistence/visibility/store/query/converter.go (hunk 1 above)
Comment (rodrigozhou) [reply common/persistence/visibility/store/query/converter.go:511]: No. index_prefixes would make search faster, but not required.

File: common/persistence/visibility/store/query/converter.go (hunk 2 above)
Comment (tdeebswihart) [reply common/persistence/visibility/store/query/converter.go:490]: Why are we removing LIKE support? Have we documented our rationale

Comment (rodrigozhou) [reply common/persistence/visibility/store/query/converter.go:490]: It doesn't behave as expected, and we don't have plans to support it in its entirety.

File: tests/advanced_visibility_test.go [hunk 3]
```diff
@@ -540,59 +540,87 @@ func (s *advancedVisibilitySuite) TestListWorkflow_KeywordQuery() {
...
-		// LIKE is supported on Elasticsearch only.
+	// Prefix search
+	listRequest = &workflowservice.ListWorkflowExecutionsRequest{
+		Namespace: s.namespace,
+		PageSize:  defaultPageSize,
+		Query:     `CustomKeywordField STARTS_WITH "justice"`,
+	}
+	resp, err = s.engine.ListWorkflowExecutions(NewContext(), listRequest)
+	s.NoError(err)
+	s.Len(resp.GetExecutions(), 1)
```
Comment (tdeebswihart) [inline tests/advanced_visibility_test.go:551]: Can you validate that you retrieve the correct record, not that we retrieve _a_ record? We could be retrieving the wrong record, but this test would still pass.

As a reviewer I'm not confident in this test

File: tests/advanced_visibility_test.go [hunk 4]
```diff
@@ -540,59 +540,87 @@ func (s *advancedVisibilitySuite) TestListWorkflow_KeywordQuery() {
...
+
+	// LIKE exact match on Keyword (supported)
+	listRequest := &workflowservice.ListWorkflowExecutionsRequest{
+		Namespace: s.namespace,
+		PageSize:  defaultPageSize,
+		Query:     `CustomKeywordField LIKE "%foo bar%"`,
+	}
+	resp, err := s.engine.ListWorkflowExecutions(NewContext(), listRequest)
+	s.NoError(err)
+	s.Len(resp.GetExecutions(), 1)
```
Comment (tdeebswihart) [inline tests/advanced_visibility_test.go:583]: Same here. Please ensure we get the expected record, not just any record

Comment (rodrigozhou) [reply tests/advanced_visibility_test.go:583]: These are old tests, soon to be deleted.

Comment (tdeebswihart) [reply tests/advanced_visibility_test.go:589]: I don't see new tests that verify this functionality works as expected, are you planning to add them?

**Review by alexshtin — APPROVED** [review]

File: tests/advanced_visibility_test.go (hunk 3 above)
Comment (rodrigozhou) [reply tests/advanced_visibility_test.go:554]: @tdeebswihart This test is new.

File: common/persistence/visibility/store/sql/query_converter_test.go [hunk 5]
```diff
@@ -328,6 +328,38 @@ func (s *queryConverterSuite) TestConvertComparisonExpr() {
...
+		{
+			name:   "not starts_with expression error",
+			input:  "AliasForKeyword01 not starts_with 123",
+			output: "",
+			err: query.NewConverterError(
+				"%s: right-hand side of '%s' must be a literal string (got: 123)",
+				query.InvalidExpressionErrMessage,
+				sqlparser.NotStartsWithStr,
+			),
+		},
```
Comment (rodrigozhou) [inline common/persistence/visibility/store/sql/query_converter_test.go:362]: @tdeebswihart New tests.

File: common/persistence/visibility/store/elasticsearch/converter_test.go [hunk 6]
```diff
@@ -93,6 +93,8 @@ var supportedWhereCases = map[string]string{
 	"create_time BETWEEN '2015-01-01 00:00:00' and '2016-02-02 00:00:00'":     `{"bool":{"filter":{"range":{"create_time":{"from":"2015-01-01 00:00:00","include_lower":true,"include_upper":true,"to":"2016-02-02 00:00:00"}}}}}`,
 	"create_time nOt between '2015-01-01 00:00:00' and '2016-02-02 00:00:00'": `{"bool":{"must_not":{"range":{"create_time":{"from":"2015-01-01 00:00:00","include_lower":true,"include_upper":true,"to":"2016-02-02 00:00:00"}}}}}`,
 	"create_time between '2015-01-01T00:00:00+0800' and '2017-01-01T00:00:00+0800' and process_id = 0 and status >= 1 and content = '三个男人' and phone = '15810324322'": `{"bool":{"filter":[{"range":{"create_time":{"from":"2015-01-01T00:00:00+0800","include_lower":true,"include_upper":true,"to":"2017-01-01T00:00:00+0800"}}},{"match":{"process_id":{"query":0}}},{"range":{"status":{"from":1,"include_lower":true,"include_upper":true,"to":null}}},{"match":{"content":{"query":"三个男人"}}},{"match":{"phone":{"query":"15810324322"}}}]}}`,
+	"value starts_with 'prefix'":     `{"bool":{"filter":{"prefix":{"value":"prefix"}}}}`,
+	"value not starts_with 'prefix'": `{"bool":{"must_not":{"prefix":{"value":"prefix"}}}}`,
```
Comment (rodrigozhou) [inline common/persistence/visibility/store/elasticsearch/converter_test.go:97]: @tdeebswihart New tests.

File: common/persistence/visibility/store/sql/query_converter_test.go (hunk 5 above)
Comment (tdeebswihart) [reply common/persistence/visibility/store/sql/query_converter_test.go:362]: I'd seen the unit tests. So we not have useful functional tests for this functionality?

File: tests/advanced_visibility_test.go [hunk 7]
```diff
@@ -540,59 +543,99 @@ func (s *advancedVisibilitySuite) TestListWorkflow_KeywordQuery() {
...
-		s.NoError(err)
-		s.Len(resp.GetExecutions(), 1)
+	listRequest = &workflowservice.ListWorkflowExecutionsRequest{
+		Namespace: s.namespace,
+		PageSize:  defaultPageSize,
+		Query:     `CustomKeywordField NOT STARTS_WITH "justice"`,
+	}
+	resp, err = s.engine.ListWorkflowExecutions(NewContext(), listRequest)
+	s.NoError(err)
+	s.Len(resp.GetExecutions(), 0)
```
Comment (rodrigozhou) [inline tests/advanced_visibility_test.go:566]: @tdeebswihart These are functional tests.

Comment (tdeebswihart) [reply tests/advanced_visibility_test.go:566]: These are the ones I raise an issue about earlier. These tests look for the presence of _a_ record (or no records) but do not verify that the returned record is what we expect.

Can you update these to ensure the returned execution actually matches the filter?I've seen tests like this (that look for the presence of _a_ record) hide bugs where the incorrect record was returned

Comment (rodrigozhou) [reply tests/advanced_visibility_test.go]: I already added checks. See L555-557.

Comment (tdeebswihart) [reply tests/advanced_visibility_test.go]: D'oh, thank you

**Pull request merged**